# Tarea 5

In [5]:
# Importrar librerías necesarias
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import RandomizedSearchCV
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

In [ ]:
# Importar funciones
from Funciones_ejer8 import train_model
from Funciones_ejer8 import predict
from Funciones_ejer8 import evaluate_model
from Funciones_ejer8 import evaluar_metricas

## Ejercicio 8

In [8]:
# Importar base de datos 
novatos = pd.read_csv("NovatosNBA.csv", sep=',')
# Eliminar columnas no numéricas o irrelevantes
novatos = novatos.drop(columns=['Unnamed: 0', 'Player', 'Team', 'Conf'])
novatos.head()

Year  Age  Games    MP   PTS   FG   FGA    FG%  3P  3PA  ...  FTpg  ORBpg  \
0  1979   22     67   726   312  122   271  0.450   0    0  ...  1.01   1.06   
1  1980   23     16   235   180   72   146  0.493   8   19  ...  1.75   0.81   
2  1979   23     82  2955  1745  693  1463  0.474  58  143  ...  3.67   2.63   
3  1980   23     13   125    56   24    52  0.462   0    0  ...  0.62   1.46   
4  1979   22     82  2027   688  275   609  0.452   2    5  ...  1.66   0.84   

   DRBpg  TRBpg  ASTpg  STLpg  BLKpg  TOVpg  PFpg  Target  
0   1.88    2.9    0.4   0.31   0.81   1.18  1.73       1  
1   1.00    1.8    1.9   0.88   0.12   1.25  1.62       1  
2   7.76   10.4    4.5   1.74   0.65   3.21  3.40       1  
3   1.54    3.0    0.2   0.31   0.15   0.62  1.92       0  
4   1.88    2.7    3.1   2.57   0.59   2.02  2.37       1  

[5 rows x 36 columns]

### Inciso a)

In [10]:
# Dividir los datos en X y y
X = novatos.drop(columns='Target').to_numpy()
y = novatos['Target'].to_numpy()

# Dividir los datos en train y test (80% entrenamiento, 20% prueba)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=42
)

In [11]:
# Primero, se genera el modelo con árboles de decisión
# Entrenar modelo
arb_dec = train_model(model=tree.DecisionTreeClassifier(),
                      X_train=X_train,
                      y_train=y_train,
                      param_distributions=dict())

# Predecir
y_pred_arb = predict(arb_dec, X_test)

# Evaluar modelo
metricas_arb = evaluate_model(arb_dec, X_test, y_test)
metricas_arb

C:\Users\sofia\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:317: UserWarning: The total space of parameters 1 is smaller than n_iter=10. Running 1 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Mejores hiperparámetros: {}
Mejor puntuación: 0.6950980297235805


{'Accuracy': 0.6942003514938488,
 'Precision': 0.6957293950873448,
 'Recall': 0.6942003514938488,
 'F1 Score': 0.6949305919076987}

La métrica de Accuracy indica que el modelo acierta en aproximadamente el 70% de los casos. En el caso de Precision, se tiene que cuando predice una clase positiva, tiene 69.6% de aciertos. Además, con Recall se ve que detecta el 69.6% de los casos positivos reales y el F1 Score muestra el promedio balanceado entre Precision y Recall, es decir, 69.6% de calidad global. Puede ser mejor.

In [17]:
# Ahora, se genera el modelo con bosques aleatorios
# Entrenar modelo
bos_ale = train_model(model=RandomForestClassifier(random_state=42),
                      X_train=X_train,
                      y_train=y_train,
                      param_distributions=dict())

# Predecir
y_pred_bos = predict(bos_ale, X_test)

# Evaluar modelo
metricas_bos = evaluate_model(bos_ale, X_test, y_test)
metricas_bos

C:\Users\sofia\anaconda3\Lib\site-packages\sklearn\model_selection\_search.py:317: UserWarning: The total space of parameters 1 is smaller than n_iter=10. Running 1 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Mejores hiperparámetros: {}
Mejor puntuación: 0.7597840925594228


{'Accuracy': 0.7908611599297012,
 'Precision': 0.7901215485774755,
 'Recall': 0.7908611599297012,
 'F1 Score': 0.7904682649896957}

Por otro lado, para el modelo de bosques aleatorios, la métrica de Accuracy indica que el modelo acierta en aproximadamente el 79% de los casos. En el caso de Precision, se tiene que cuando predice una clase positiva, tiene 79% de aciertos. Además, con Recall se ve que detecta el 79% de los casos positivos reales y el F1 Score muestra el promedio balanceado entre Precision y Recall, es decir, 79% de calidad global. Parecer ejercer un mejor trabajo que el modelo de árboles de decisión.

### Inciso b)

In [ ]:
# DataFrame comparando modelos actuales con los de tareas anteriores, usando metricas
resultados = []

# Árboles de decisión
resultados.append(evaluar_metricas("Árboles de decisión", y_test, y_pred_arb))

# Bosques aleatorios
resultados.append(evaluar_metricas("Bosques aleatorios", y_test, y_pred_bos))

# Se va a comparar con LDA, QDA y Naive Bayes. Por lo tanto, hay que calcular las métricas
from Funciones_ejer8 import correr_ejer_4
y_test, lda_pred, qda_pred, naive_pred = correr_ejer_4()

# Obtener predicciones
# Evaluar
resultados.append(evaluar_metricas("LDA", y_test, lda_pred))
resultados.append(evaluar_metricas("QDA", y_test, qda_pred))
resultados.append(evaluar_metricas("Naive Bayes", y_test, naive_pred))

df_resultados = pd.DataFrame(resultados)
df_resultados

C:\Users\sofia\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:1024: LinAlgWarning: The covariance matrix of class 0 is not full rank. Increasing the value of parameter `reg_param` might help reducing the collinearity.
  warnings.warn(
C:\Users\sofia\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:1024: LinAlgWarning: The covariance matrix of class 1 is not full rank. Increasing the value of parameter `reg_param` might help reducing the collinearity.
  warnings.warn(


Modelo  Precisión global  Error global      PP      PN
0  Árboles de decisión            0.6957        0.3043  0.7703  0.5528
1   Bosques aleatorios            0.7901        0.2099  0.8382  0.6979
2                  LDA            0.7836        0.2164  0.8219  0.7102
3                  QDA            0.7728        0.2272  0.8949  0.5385
4          Naive Bayes            0.7740        0.2260  0.9091  0.5148

QDA y Naive Bayes tienen los PP más altos, sin embargo, también poseen el PN más bajo. Por otro lado, Bosques Aleatorios cuenta con la precisión global más alta y, por tanto, el error global más bajo. Además, a pesar de no tener los mejores PP o PN, ambos son relativamente altos. Por tanto, se concluye que el modelo de Bosques Aleatorios es el mejor.

## Ejercicio 9